### ISSUE TITLE
Getting invocations of passed variables

### ISSUE THREAD

**Issue by mohammadnaseri** [INRIA/spoon#1754]
Hi, 

I would like to know if Spoon supports a way to find all the invocations of a variable, not necessarily in current scope, but also in the methods that the variable was passed. 
If not, what is the best way for the problem?

Thanks,

**Comment by pvojtechovsky** [comment]
Hi and welcome ;-)

There is `VariableReferenceFunction` for that, which searches all references to any variable (Local variable, Parameter, CatchVariable) in whole scope of it's visibility (including nested class and lambda expressions).

> not necessarily in current scope, but also in the methods that the variable was passed.

Do you mean something like this code?
```java
void method1() {
  int var1 = 7;
  method2(var1); //this reference is found by VariableReferenceFunction
  System.out.println(var1) //this reference is found by VariableReferenceFunction
}

void method2(int param1) {
  System.out.println(param1) //this reference is NOT found by VariableReferenceFunction
}
```
If you want to find `param1`, then you need to analyze a data flow/control flow. Spoon has nothing for that yet. There is an old project https://github.com/marcelinorc/spooncontrolflow, which started with that...

**Comment by mohammadnaseri** [comment]
Thanks for the answer, 
So in the following case, 

``` 
void method1() {
     myClass var = new myClass();
     var.testMethod1();
     method2(var);
}

void method2(myClass var) {
    var.testMethod2();
    method3(var);
}

void method3(myClass var) {
    var.testMethod3();
}
 ```
if I want to retrieve both ```testMethod1()``` , ```testMethod2()``` and ```testMethod3()```, I have to track the flow myself?
In other words, I want to find all the methods that are being called on the specific object.

**Comment by pvojtechovsky** [comment]
yes, you have to do it yourself, but there are good primitive tools (CtQuery) in Spoon, which will support you with that analyzing algorithm.

**Comment by surli** [comment]
ping @mohammadnaseri did you find a solution for your problem then?

**Comment by mohammadnaseri** [comment]
@surli Yes, As I couldn't use CtQuery, so I am tracking the flow recursively.
